### ISSUE TITLE
WebKitWebProcess stucks with luakit.process_limit = 1

### ISSUE THREAD

**Issue by bodqhrohro** [luakit/luakit#754]
## I'm submitting a…
Bug Report

## Current Behavior
With this setting, no sites are loaded: some stuck on some load percent, some just freeze, but in general, nothing is displayed in the viewport. WebKitWebProcess appears, it occupies 120M RAM and has an S state, but produces no CPU activity.

## Expected Behavior
Epiphany has a similar `org.gnome.epiphany.process-model = 'shared-secondary-process'` setting, which works well, despite of being abandoned. So I expect LuaKit to work in single-process mode too.

## Reproduction Instructions
Set `luakit.process_limit = 1` in `rc.lua`.

## Environment

Linux Distribution & Version: Debian 10
Output of `luakit --version`: luakit 2.0.0
  built with webkit 2.22.5 (installed version: 2.22.5)

**Comment by aidanholm** [comment]
I can't reproduce this; luakit runs fine for me with only one web process, with webkit 2.22.6.

**Comment by bodqhrohro** [comment]
Yeah, I've got the more exact reason: WebKitNetworkProcess eventually hangs and starts consuming the whole CPU core, thus loading stucks everywhere. If I kill it manually, the loading usually resumes automatically, sometimes it fails with an "internal error".

Such hangs happen with process_limit=2 too, but more rarely than with process_limit=1. They happen both with enabled and disabled proxy.

**Comment by bodqhrohro** [comment]
I found a reliable way to make WebKitNetworkProcess hang when using a proxy: visiting a non-existent domain (f.e., http://robovm.com/). Without a proxy it just fails, but with a SOCKS5 proxy it hangs, probably because of remote DNS.

**Comment by c0dev0id** [comment]
Does this still happen? There were significant changes in Webkit 2.26 related to how processes work. I think process_limit is ignored meanwhile. Are you able to reproduce the issue with a new webkit version?

**Comment by bodqhrohro** [comment]
The process limit was removed in 2.26, as well as the support for GTK+2 NPAPI plugins That's not suitable for me, so I just have frozen 2.24.

WebKitGTK+ has a single-process mode, which still can be enabled in Epiphany with the /org/gnome/epiphany/process-model gsetting. Though LuaKit had strictly migrated to the multi-process mode several years ago, and since that, the codebase changes rely on it, so I couldn't build&run the fresh LuaKit successfully just by changing the mode back.

**Comment by c0dev0id** [comment]
Interesting. I didn't know that this mode still exists. I'll play with it. Maybe we can make it work. But NPAPI support will certainly never come back.

**Comment by c0dev0id** [comment]
Starting luakit with `WEBKIT_USE_SINGLE_WEB_PROCESS=1` set in the environment works fine for me. It spawns one WebKit process for all tabs and I haven't seen an issue in the last hour. I'll keep this set for now.

**Comment by c0dev0id** [comment]
I think we can close this. Or is there anything to add in this discussion? Feel free to reopen.

**Comment by bodqhrohro** [comment]
Actually, the caption is misleading, as the issue turned out to be this: https://github.com/luakit/luakit/issues/754#issuecomment-468711047 I did not notice that when you first resurrected the issue and started discussing replacements to the process limit which I successfully use since that.

If the proxy issue is still relevant, it should be opened separately, though I barely will switch to upstream to test it anymore.

**Comment by c0dev0id** [comment]
Okay, then I'm leaving this open. I'm playing with the proxy module anyway... I'll try to make it suck while at it...

**Comment by c0dev0id** [comment]
I've retested the proxy issue and I can't reproduce it. When I configure a proxy and visit a domain that doesn't exist I get reliably:
```
Unable to load page

A problem occurred while loading the URL http://robovm.com/

Connection through SOCKSv4 server was rejected
```

However, this was a socks4 proxy. What socks5 proxy did you use? I think webkit waits for the proxy to respond somehow. And if it doesn't it may hang.

**Comment by bodqhrohro** [comment]
>and visit a domain

It's better reproduced with pages that do AJAX requests in background, and were loaded with a proxy disabled, but try to make new requests via a proxy. For example, the Syncthing Web GUI which makes frequent requests to localhost (which obviously is not available through a proxy).

>What socks5 proxy did you use?

Usually a SSH tunnel to my VPS (`ssh -CND <port> ...`, and `socks5://127.0.0.1:<port>` in LuaKit), or TOR.(`socks5://127.0.0.1:9050`).
